File: benchmarks/semantic_search/fixtures/go-workflow-discovery-v1/internal/selection/validator.go
package selection

import (
	"fmt"

	"example.com/synthetic-workflow-discovery/internal/discovery"
)

type ValidationError struct {
	WorkflowID string
	Reason     string
}

func (e *ValidationError) Error() string {
	return fmt.Sprintf("workflow %q rejected: %s", e.WorkflowID, e.Reason)
}

type Validator struct {
	allowed map[string]struct{}
	state   *discovery.WorkflowState
}

func NewValidator(allowedWorkflowIDs []string) *Validator {
	allowed := make(map[string]struct{}, len(allowedWorkflowIDs))
	for _, id := range allowedWorkflowIDs {
		allowed[id] = struct{}{}
	}
	return &Validator{allowed: allowed}
}

func (v *Validator) SetWorkflowState(state *discovery.WorkflowState) {
	v.state = state
}

// IsAllowed requires catalog membership and, when state is attached, current
// list_workflows membership.
func (v *Validator) IsAllowed(workflowID string) bool {
	if _, ok := v.allowed[workflowID]; !ok {
		return false
	}
	return v.state == nil || v.state.Contains(workflowID)
}

func (v *Validator) Validate(workflowID string) error {
	if v.IsAllowed(workflowID) {
		return nil
	}
	return &ValidationError{WorkflowID: workflowID, Reason: "not in current discovery result"}
}

File: benchmarks/semantic_search/fixtures/rust-workflow-discovery-v1/src/selection/validator.rs
use crate::discovery::state::WorkflowState;

#[derive(Debug)]
pub struct ValidationError {
    pub workflow_id: String,
}

pub struct WorkflowValidator {
    allowed_ids: std::collections::HashSet<String>,
    workflow_state: Option<WorkflowState>,
}

impl WorkflowValidator {
    pub fn new(allowed_ids: std::collections::HashSet<String>) -> Self {
        Self {
            allowed_ids,
            workflow_state: None,
        }
    }

    pub fn set_workflow_state(&mut self, state: WorkflowState) {
        self.workflow_state = Some(state);
    }

    pub fn is_allowed(&self, workflow_id: &str) -> bool {
        self.allowed_ids.contains(workflow_id)
            && self
                .workflow_state
                .as_ref()
                .is_none_or(|state| state.contains(workflow_id))
    }

    pub fn validate(&self, workflow_id: &str) -> Result<(), ValidationError> {
        if self.is_allowed(workflow_id) {
            Ok(())
        } else {
            Err(ValidationError {
                workflow_id: workflow_id.to_owned(),
            })
        }
    }
}
